src/include/mpir_hw_topo.h

```diff
@@ -0,0 +1,184 @@
+#define MPIR_Network_edge void
+
+/*
+ * Hardware topology function table
+ */
+typedef struct {
+    /*
+     * Return logical ids for object type covering current PU
+     *
+     * in parameters
+     *  obj_type: valid MPIR_Node_hw_obj_type
+     *
+     * out parameters
+     *  size:     number of returned node objects (len of obj_ids)
+     *  obj_ids:  list of object type logical identifiers
+     *
+     * return value
+     *  0 no errors
+     * -1 otherwise
+     *
+     * errors
+     *  **node_error_object: invalid object type
+     *  **node_error_other:
+     */
+    int (*node_get_covering_obj) (MPIR_Node_hw_obj_type obj_type, int *size, int **obj_ids);
+
+    /*
+     * Return network topology type
+     *
+     * TODO: add detailed description
+     */
+     MPIR_Network_type(*network_get_type) ();
+
+    /*
+     * Return network topology node type
+     *
+     * TODO: add detailed description
+     */
+     MPIR_Network_node_type(*network_get_node_type) (MPIR_Network_node * node);
+
+    /*
+     * Return a pointer to this node in network topology
+     *
+     * TODO: add detailed description
+     */
+    MPIR_Network_node *(*network_get_endpoint) ();
+
+    /*
+     * Return edge destination node
+     *
+     * TODO: add detailed description
+     */
+    MPIR_Network_node *(*network_get_edge_dest_node) (MPIR_Network_edge * edge);
+
+    /*
+     * Return number of nodes in network topology
+     *
+     * TODO: add detailed description
+     */
+    int (*network_get_num_nodes) ();
+
+    /*
+     * Return the number of edges and edge pointers for this node in network topology
+     *
+     * TODO: add detailed description
+     */
+    int (*network_get_all_edges) (int *num_edges, MPIR_Network_edge *** edges);
+
+    /*
+     * Return dimension of the torus network topology
+     *
+     * TODO: add detailed description
+     */
+    int (*network_torus_get_dimension) ();
+
+    /*
+     * Return geometry of the torus network topology
+     *
+     * TODO: add detailed description
+     */
+    int *(*network_torus_get_geometry) ();
+
+    /*
+     * Return node index in the torus network topology
+     *
+     * TODO: add detailed description
+     */
+    int (*network_torus_get_node_index) ();
+
+    /*
+     * Return node level in tree network topology
+     *
+     * TODO: add detailed description
+     */
+    int *(*network_tree_get_node_levels) ();
+
+    /*
+     * Return the hostnode index in the tree network topology
+     *
+     * TODO: add detailed description
+     */
+    int (*network_tree_get_hostnode_index) (int *node_index, int *num_nodes);
+
+    /*
+     * Return list of switches at certain level in the tree
+     *
+     * TODO: add detailed description
+     */
+    int (*network_tree_get_switches_at_level) (int switch_level,
+                                               MPIR_Network_node *** switches_at_level,
+                                               int *switch_count);
```

Comment:
Functions specific to a topology should be marked as such inside a separate structure.  So, `tree` functions should go in a structure.  `torus` functions should go in a different structure.  These structures can be in a union.  That way, it's clear that only one set is valid at a time.

Reply:
Doesn't that have a lot of levels of dereference? Following this logic I should probably have a `node_struct` for node functions and then a `network_struct` for network functions. Then inside `network_struct` I can have the `union`. Thus accessing topology function would look like `hw_topo_fns->node->get_processor_obj`, `hw_topo_fns->network->get_type` and `hw_topo_fns->network->u.tree->get_switches_at_level`.

Reply:
Structures are not dereferences.  `hw_topo_fns->node.get_processor_obj`, `hw_topo_fns->network.u.tree.get_switches_at_level`.